src/DataStructures/Variables.hpp

```diff
@@ -711,26 +697,21 @@ Variables<tmpl::list<Tags...>>& Variables<tmpl::list<Tags...>>::operator=(
 
 /// \cond HIDDEN_SYMBOLS
 template <typename... Tags>
-template <typename TagToAdd, typename... Rest,
-          Requires<tt::is_a<Tensor, typename TagToAdd::type>::value>>
-void Variables<tmpl::list<Tags...>>::add_reference_variable_data(
-    tmpl::list<TagToAdd, Rest...> /*unused*/,
-    const size_t variable_offset) noexcept {
-  ASSERT(size_ > (variable_offset + TagToAdd::type::size() - 1) *
-                     number_of_grid_points_,
-         "This ASSERT is typically triggered because a Variables class was "
-         "default constructed. The only reason the Variables class has a "
-         "default constructor is because Charm++ uses it, you are not "
-         "supposed to use it otherwise.");
-  typename TagToAdd::type& var =
-      tuples::get<TagToAdd>(reference_variable_data_);
-  for (size_t i = 0; i < TagToAdd::type::size(); ++i) {
-    var[i].set_data_ref(
-        &variable_data_[(variable_offset + i) * number_of_grid_points_],
-        number_of_grid_points_);
-  }
-  add_reference_variable_data(tmpl::list<Rest...>{},
-                              variable_offset + TagToAdd::type::size());
+void Variables<tmpl::list<Tags...>>::add_reference_variable_data() noexcept {
+  if (size_ == 0) {
+    return;
+  }
+  variable_data_.reset(variable_data_impl_.get(), size_);
+  size_t variable_offset = 0;
+  tmpl::for_each<tags_list>([this, &variable_offset](auto tag_v) noexcept {
```

Comment:
Please don't use `tmpl::for_each` here. The loop with pointer arithmetic was intentional because at least when I did timings it was a lot faster.

Reply:
I didn't remove the loop with pointer arithmetic.  It's three lines below here.

Reply:
Oh, I see now. Sorry, I misread before. You're replacing the manual pack expansion with a `tmpl::for_each`. Great, thanks!